### PR TITLE
Fix Resource version properties lookup and extract version file gener…

### PR DESCRIPTION
**Pull request by anuraaga**
…ation.

I think it's good practice to add the version to a file in any JARs we publish as it allows verification of version alignment at runtime as needed. So this extracts creation of the version specifier to the java conventions itself. Note that `MANIFEST` tends to not work well for this because fatjars often lose it.

### CODE REVIEW

File: sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java [hunk 1]
```diff
@@ -211,7 +211,8 @@ void testDefaultResources() {
     assertThat(attributes.get(ResourceAttributes.SERVICE_NAME)).isEqualTo("unknown_service:java");
     assertThat(attributes.get(ResourceAttributes.TELEMETRY_SDK_NAME)).isEqualTo("opentelemetry");
     assertThat(attributes.get(ResourceAttributes.TELEMETRY_SDK_LANGUAGE)).isEqualTo("java");
-    assertThat(attributes.get(ResourceAttributes.TELEMETRY_SDK_VERSION)).isNotNull();
```
Comment (anuraaga) [inline sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java:214]: This used to pass since it was `"unknown"` 😱

**Comment by Oberon00** [comment]
I don't think we want/need this version.properties everywhere. Please take a look at #3461 where I only generate it into the sdk-common jar. Note that only one of #3461 or this PR can be merged.

**Comment by anuraaga** [comment]
@Oberon00 the version files can be used to verify version alignment and have almost no impact on artifact size. Why not want them everywhere?

**Comment by Oberon00** [comment]
So, I like the additional changes of this PR (looking in sdk/common/ instead of sdk/, additional tests) but I don't see why we need the version.properties in every jar. We have the Implementation-Version in the MANIFEST.MF in every jar already. As suggested in https://github.com/open-telemetry/opentelemetry-java/pull/3461#issuecomment-896070884, we could use the manifest version directly instead of generating *any* version.properties.

**Comment by anuraaga** [comment]
Yeah but

> Note that MANIFEST tends to not work well for this because fatjars often lose it.

In my experience, MANIFEST tends to be mostly useless in practice since someone will file a bug when their fatjar causes problems with it.

**Comment by Oberon00** [comment]
OK, that's a valid argument. It *is* potentially useful to have the version.properties in the API jar at least.

But I'd add the version.properties to the publishing conventions instead of the java conventions, or do we also want these for unpublished jars?

**Comment by anuraaga** [comment]
@Oberon00 Agree that publish conventions would be enough, but the Gradle configuration was simpler in `java-conventions` with the availability of the `moduleName`. The particular issue is that we have `bom` artifacts that are published, but are not `java-library`. The downside for non-published is just a very small amount of build time I think so figured it's OK.

**Comment by Oberon00** [comment]
How about mixing in the old approach and using `plugins.withId("maven-publish")` in the java conventions?

**Comment by anuraaga** [comment]
Thanks, nice idea!

**Comment by Oberon00** [comment]
Actually, instead of `maven-publish`, maybe using the ID of the otel publishing convention plugin would be better?

File: buildSrc/src/main/kotlin/otel.java-conventions.gradle.kts [hunk 2]
```diff
@@ -136,6 +136,29 @@ tasks {
   }
 }
 
+// Add version information to published artifacts.
+plugins.withId("maven-publish") {
+  tasks {
+    register("generateVersionResource") {
+      val moduleName = otelJava.moduleName
+      val propertiesDir = moduleName.map { file("build/generated/properties/${it.replace('.', '/')}") }
```
Comment (Oberon00) [inline buildSrc/src/main/kotlin/otel.java-conventions.gradle.kts:144]:
```suggestion
      val propertiesDir = moduleName.map { File(buildDir, "generated/properties/${it.replace('.', '/')}") }
```

File: buildSrc/src/main/kotlin/otel.java-conventions.gradle.kts [hunk 3]
```diff
@@ -136,6 +136,29 @@ tasks {
...
+
+      doLast {
+        File(propertiesDir.get(), "version.properties").writeText("sdk.version=${project.version}")
+      }
+    }
+  }
+
+  sourceSets {
+    main {
+      output.dir("build/generated/properties", "builtBy" to "generateVersionResource")
```
Comment (Oberon00) [inline buildSrc/src/main/kotlin/otel.java-conventions.gradle.kts:157]:
```suggestion
      output.dir("$buildDir/generated/properties", "builtBy" to "generateVersionResource")
```

**Review by Oberon00 — APPROVED** [review]

**Review by jkwatson — APPROVED** [review]
Thanks!

**Pull request merged**